Fixed computation of intersectsRay() for Cylinder, Box and ConvexMesh

peci1 opened the pull request:
This PR fixes a multitude of problems the current implementations of `intersectsRay()` had.

In Box, even according to the link to the used method, the approach should only work for axis-aligned boxes. But bodies::Box can be arbitrarily transformed. So I instead strip rotation from both the box and the ray, which gives the same results. The old implementation would fail every time the untransformed min corner ceased to be the min corner of the transformed box.

For Cylinder, there were a few cornercases when incorrect number of intersections were returned (e.g. when the ray is tangential to the cylinder or when it goes through the edge between base and sides).

For ConvexMesh, there were several problems. Many of them arose from a wrong approach to incorporate padding. Also, if a ray hit the edge of two triangles, it was reported twice.

I also restructured the tests for both `containsPoint()` and `intersectsRay()` and added a lot of new tests - both randomized and testing corner cases. There is also a regression test `BoxRayIntersection::Regression109` which fails on current melodic-devel and succeeds with this test.

Aside from the fixes in `intersectsRay()`, I also fixed `Cylinder::samplePointInside()` and `ConvexMesh::containsPoint()` whose correct working is needed by the unit tests.

peci1 commented:
This would deserve a few tests, they're coming tomorrow. For the time being I verified the algorithm in RViz.

rhaschke commented on src/bodies.cpp:
```diff
@@ -595,34 +595,40 @@ void bodies::Box::computeBoundingCylinder(BoundingCylinder& cylinder) const
   cylinder.radius = sqrt(a * a + b * b);
 }
 
+// for some reason, the O2 optimization screws up this function which then gives wrong results
```
What?

peci1 replied:
No idea... That's what I observed. Turning off O2 optimization solved the issue.

rhaschke commented on src/bodies.cpp:
```diff
@@ -535,7 +535,7 @@ void bodies::Box::updateInternalData()
   normalH_ = basis.col(2);
 
-  const Eigen::Vector3d tmp(normalL_ * length2_ + normalW_ * width2_ + normalH_ * height2_);
+  const Eigen::Vector3d tmp(length2_, width2_, height2_);
```
Obviously this changes the semantics of `corner1_` and `corner2_`. Could you please document this in the header? Does this change have implications for other pieces of the code?

Could we remove the helper variables normalL, normalW, normalH?
These are only used in Box::containsPoint(), which could be simplified to:
```
const Eigen::Vector3d aligned = (pose_.rotation().transpose() * (p - center_)).abs();
return aligned[0] <= length2_ && aligned[1] <= width2_ && aligned[2] <= height2_;
```

peci1 replied:
Ok, I'll have a look at this.

peci1 replied:
I think you definitely don't want to call `rotation()` in `containsPoint` as this function calls SVD if I'm not wrong. The question is if calling `linear()` wouldn't do the same job here (which I think would).

Changing the semantics of `corner1_` and `corner2_` wouldn't change anything in this library except for the ray-box test. The variables are protected, so I think chances that somebody else would be using them is low. But maybe it'd be better to rename them so that people who were using these variables will get a build failure. Or I can leave the old variables there and add new ones, but as the old ones were semantically wrong, I'm more for the option of getting rid of them.

rhaschke replied:
> I think you definitely don't want to call rotation() in containsPoint as this function calls SVD if I'm not wrong.

It was calling SVD, when we used Affine3d, but since we use Isometry3d, Eigen knows thatit can just return the 3x3 rotation sub-matrix. No overhead at all.

> Changing the semantics of corner1_ and corner2_ wouldn't change anything in this library.

I just wanted to be sure. No need to change the variable names. Just document their semantics in the header (there was nothing before).

rhaschke reviewed (dismissed):
Generally, this looks good to me.
Please provide a description for this PR, mentioning that this depends on other PRs.
Please also provide the promised unit tests (one of them should be failing before this fix).

peci1 committed:
Fixed computation of the ray-box intersections.

Even according to the link to the used method, the approach should only
work for axis-aligned boxes. But bodies::Box can be arbitrarily transformed.
So I instead strip rotation from both the box and the ray, which gives
the same results.

The old implementation would fail every time the untransformed min corner
ceased to be the min corner of the transformed box.

# Conflicts:
#	src/bodies.cpp

peci1 commented:
I added PR description and separated this PR from the others as it doesn't depend on them semantically. I'll have a look at the tests.

peci1 committed:
Moved all ray intersection tests to a dedicated test file.

peci1 committed:
Refactored intersection tests for better readability.

peci1 committed:
Added more intersectsRay() tests.

peci1 committed:
Fixed computation of Box::intersectsRay() in case the ray originates within the box.

A false second intersection was returned in that case, but as the ray starts inside the box, there should only be one intersection.

peci1 committed:
Simplified Box::containsPoint() and removed unnecessary helper variables.

peci1 committed:
Minor improvement in Box::intersectsRay().

peci1 committed:
Fixed ConvexMesh::containsPoint() to work correctly with padding.

peci1 committed:
Fixed Cylinder::samplePointInside() not transforming the generated point by the transform of the body.

peci1 committed:
Fixed Cylinder::intersectsRay() to correctly return one intersection in case the ray is touching the cylinder side.

Also fixed the case when the ray goes exactly at the edge between base and sides.

peci1 committed:
Fixed a case which could result in unwanted NaNs in Box::intersectsRay().

peci1 committed:
Stopped preventing GCC optimizations on Box::intersectsRay().

The reported problems no longer occur.

peci1 committed:
Fixed computation of ConvexMesh::intersectsRay() for padded meshes and some other bugs.

peci1 committed:
Prevent reporting identical intersections multiple times in ConvexMesh::intersectsRay().

peci1 committed:
Added more containsPoint() tests.

peci1 committed:
Kinetic compatibility.

peci1 commented:
I had a look at this again, rebased on current melodic-devel, and added a lot of tests. First, I only added box-ray tests, but then I decided to add tests for all bodies. And along the way, I've discovered and fixed a lot of other bugs in various more or less related functions, like `containsPoint()` and `samplePoint()`. Check out the updated title and description of this PR.

I would've sent these other fixes as another PRs, but the tests in this one would fail without them. But if it's desired, I can take the effort and split this into more PRs - I just do not want to take this effort with the tests, so I'd separate them as the "final" PR, and the prior ones would come without tests.

I did some API/ABI breaks here. 

1. First, as requested by @rhaschke , I removed the unnecessary internal fields `normalL_`, `normalW_` and `normalH_` of `bodies::Box`. I instead added `invPose_`.
1. When I was already breaking API/ABI, I also changed `MeshData` by adding one more field.
1. Not really API/ABI break, but I changed the meaning of `corner1_` and `corner2_` protected members of `bodies::Box`. It should not affect anything but potential user code that would access these fields. But as they had no docs before, I think the chance there is some other code using them is pretty minimal.

If API/ABI breaks are not desired for melodic-devel, 1) can be avoided completely (and the API break be left to noetic), and 2) can also be worked around somehow (last resort would be the translation-unit-private global variable ugly trick).

For some reason, CI did not trigger here (maybe because of the pending requested changes?), but here's mine which shows okay: https://travis-ci.com/github/peci1/geometric_shapes . I've also run a local test with this PR and the whole Moveit repo. It ran fine, except for a few `ushort_test/MeshFilterTestUnsignedShort.unsigned_short` tests, which however fail for me even on melodic-devel branch of geometric_shapes.

As with my other PRs, this may of course break user code that counted with the buggy behavior. But (thanks to the tests), I'm pretty sure that if the library is used correctly, it should not bring any new problems.

v4hn commented on src/bodies.cpp:
```diff
@@ -454,9 +454,9 @@ bool bodies::Cylinder::intersectsRay(const Eigen::Vector3d& origin, const Eigen:
+    if (d >= 0.0 && fabs(a) > detail::ZERO)
     {
-      d = sqrt(d);
+      d = (d != 0.0 ? sqrt(d) : 0.0);
```
Why would you special-case 0?

v4hn replied:
In another pull-request you started using `Eigen::numexpt::sqrt`.
Are there reasons to stick to one of them consistantly?

If we would need the `#ifdef` for the eigen version everywhere, that would be a reason to leave it out or add it through auxiliary headers...

peci1 replied:
You're right, the statement is weird and probably functionally equivalent to what was there before. I'll revert this line.

peci1 replied:
Regarding `Eigen::numext::sqrt`: this is from its docs:

> It is essentially equivalent to `using std::sqrt; return sqrt(x);`,  but slightly faster for float/double and some compilers (e.g., gcc), thanks to specializations when SSE is enabled.

So I just thought the compiler could get a chance to use some faster instructions in some cases. But I agree that the ifdefs needed for kinetic would really uglify the code

v4hn replied:
Personally, I'm not fully convinced the vectorized sqrt actually make a noticeable difference.
As always, if going for a more unusual construct, benchmark it's actually faster.
Please leave it out for now and maybe later provide a patch to change all calls if benchmarks show it's actually worthwhile.

v4hn commented on src/bodies.cpp:
```diff
@@ -634,34 +633,47 @@ void bodies::Box::computeBoundingBox(bodies::AABB& bbox) const
   bbox.extendWithTransformedBox(getPose(), 2 * Eigen::Vector3d(length2_, width2_, height2_));
 }
 
+// for some reason, the O2 optimization screws up this function which then gives wrong results
```
You removed the actual `GCC` directives again without further comments, but left the comment saying it's broken.

Please either (1) find the gcc bug (report) and link it here, (2) remove the comment if it's not a problem, (3) add the compiler version you used and reference a concrete test that fails.

peci1 replied:
I was unable to reproduce the bug in recent tests. So it probably "got resolved". I'll delete the comment.

v4hn commented on src/bodies.cpp:
```diff
@@ -699,9 +711,16 @@ bool bodies::Box::intersectsRay(const Eigen::Vector3d& origin, const Eigen::Vect
+      }
+      else
+      {
         intersections->push_back(tmax * dir + origin);
```
Does `tmin <= detail::ZERO` really imply origin is inside the box?
If it does, please add a comment to make this clear here.

peci1 replied:
That's it. I added comments to the code which should explain everything.

v4hn replied:
wonderful, thanks! :sun_with_face:

v4hn reviewed (approved):
Please address the inline comments. Aside from them, I approve.

v4hn commented:
Lastly, this patch conflicted with one of the others I merged. Could you please rebase?

peci1 commented:
I've made the requested changes.

peci1 commented:
Thanks for the reviews, @v4hn ;)

v4hn commented:
On Fri, Apr 24, 2020 at 10:17:21AM -0700, Martin Pecka wrote:
> Thanks for the reviews, @v4hn ;)

Thanks for your continuous effort with these patches!
I still fail to understand how all these issues could ever have made it through the years...

peci1 commented:
I think insisting on good tests for PRs is the way to prevent most of such errors for the future. I've seen moveit repos already report some kind of test coverage... What about this repo? Is it planned?

v4hn commented:
https://github.com/ros-planning/geometric_shapes/pull/120 :-)

rhaschke commented on include/geometric_shapes/bodies.h:
```diff
@@ -384,12 +384,10 @@ class Box : public Body
 
-  Eigen::Vector3d corner1_;
-  Eigen::Vector3d corner2_;
+  Eigen::Vector3d corner1_;  //!< The translated, but not rotated min corner
```
What about renaming these vectors to min_corner_ and max_corner_?
As you renamed `normal[L,W,H]` anyway, ABI and API is broken anyway and we need to target this to Noetic only.

peci1 replied:
Okay, I can rename them for noetic. As I'd still like to see this in melodic, should I split this PR into two and try to make this one compatible with melodic? In my view, this approach would be easier than first merging to noetic and then backporting...

As discussed earlier, the API/ABI breaks can be avoided at the cost of minor uglification of code.

rhaschke replied:
To maintain ABI you would need to keep the axis vectors instead of `invRot_`. More ugly, but ok.
The other addition I have seen is `std::map<unsigned int, unsigned int> triangle_for_plane_`.
Didn't look for what it is actually used. Replacing this by a global variable, I would actually *not* support. Why is it needed now? Just to be faster?

peci1 replied:
Walking on a thin ice, I guess a `Matrix3d` could take up the same amount of memory as three `Vector3d`s. But that could run in many issues once memory alignment comes into play. So I rather give up this idea.

`invRot` would have to be recomputed every time it is used, but as it is only a transpose, it shouldn't be that bad in melodic. I can also imagine storing the rows/cols of the inverted matrix in the abandoned vectors, but I bet the performance could be even worse than transposing it where needed.

The `triangle_for_plane` is really required for efficiency - otherwise mesh-ray intersection tests would need to re-enumrate over all triangles to find which plane the current one belongs to. The global variable trick is something Gazebo people are willing to accept if it is just "temporary" to get a feature to a released version; the development version then implements the nicer, but ABI-breaking version. An example is here: https://github.com/osrf/gazebo/commit/415be4050e955129956084bcb7bb4c2b436cd943 . It creates a global map, which stores references to created objects (ConvexMesh in this case) and maps them to the structure that is needed to be added. Ugly, but works.

Actually, I'd consider making the whole `MeshData` PIMPL for noetic (i.e. moving it to the .cpp file), so that storing additional data will not cause an ABI break any more. The same could be done for the other bodies, too, but there would be the additional cost of dynamic memory allocation. The ConvexMesh allocates memory anyways, so adding one more unique_ptr should not be a problem.

peci1 replied:
I implemented ABI compatibility using the global variable.

rhaschke commented on include/geometric_shapes/bodies.h:
```diff
@@ -458,7 +456,12 @@ class ConvexMesh : public Body
+  /** \brief Check if the point is inside all halfspaces this mesh consists of (mesh_data_->planes_).
+   *
+   * \note The point is expected to have pose_ "cancelled" (have inverse pose of this mesh applied to it)
+   *       and it is expected to be scaled and padded.
```
What does it mean for a point to be scaled and padded?

peci1 replied:
I edited the documentation.

rhaschke commented on src/bodies.cpp:
```diff
@@ -497,8 +497,18 @@ bool bodies::Cylinder::intersectsRay(const Eigen::Vector3d& origin, const Eigen:
+  {
+    if (intersections->size() == n)
+      break;
+    if (!intersections->empty() && (p.pt - intersections->back()).squaredNorm() < pow(detail::ZERO, 2))
```
What about using Eigen's `isApprox()` function here?

peci1 replied:
Good catch, it looks much better.

rhaschke commented on src/bodies.cpp:
```diff
@@ -513,17 +523,8 @@ bool bodies::Box::samplePointInside(random_numbers::RandomNumberGenerator& rng,
-
-  double pH = v.dot(normalH_);
-  return fabs(pH) <= height2_;
+  const Eigen::Vector3d aligned = (invRot_ * (p - center_)).cwiseAbs();
```
Great simplification! Much more readible! :+1:

rhaschke commented on src/bodies.cpp:
```diff
@@ -640,28 +639,45 @@ bool bodies::Box::intersectsRay(const Eigen::Vector3d& origin, const Eigen::Vect
+  // and a final constraint on tmin and tmax is updated by each of these projections. If tmin > tmax, there is no
+  // intersection between the line and the box.
+
+  divx = 1 / d.x();
```
The following code seems to be repeating the same pattern for x,y, and z.
Wouldn't it make sense to use a for loop from 0..2 instead and store the results min/max[x,y,z] in vectors?
Probably it's even possible to vectorize the whole operation?

peci1 replied:
I'll have a look at it. The idea was early stopping. But if vectorization could come into play, it might be faster than the early stopping. There will just be more problems handling nans...

rhaschke replied:
As far as I have seen, you are computing the same expression, componentwise.
Hence, you could do the nan-test componentwise as well. That's not more complicated, is it?

peci1 replied:
I think I found a very nice vectorization and simplification of the algorithm. The tests pass, so it should be correct.

rhaschke commented on src/bodies.cpp:
```diff
@@ -716,8 +748,8 @@ bool bodies::ConvexMesh::containsPoint(const Eigen::Vector3d& p, bool /* verbose
   {
+    // Transform the point to the "base space" of this mesh
     Eigen::Vector3d ip(i_pose_ * p);
-    ip = mesh_data_->mesh_center_ + (ip - mesh_data_->mesh_center_) / scale_;
```
What a weird code!?!

rhaschke commented on src/bodies.cpp:
```diff
@@ -1002,7 +1035,7 @@ void bodies::ConvexMesh::updateInternalData()
       new shapes::Box(mesh_data_->box_size_.x(), mesh_data_->box_size_.y(), mesh_data_->box_size_.z()));
   bounding_box_.setDimensions(box_shape.get());
```
Use the set*Dirty() functions here?

peci1 replied:
Done.

rhaschke commented on src/bodies.cpp:
```diff
@@ -1094,7 +1127,13 @@ bool bodies::ConvexMesh::isPointInsidePlanes(const Eigen::Vector3d& point) const
+    const auto scaled_point_on_plane =
+        scaled_vertices_->at(mesh_data_->triangles_[3 * mesh_data_->triangle_for_plane_[i]]);
+    const double w_scaled_padded = -plane_vec.dot(scaled_point_on_plane);
+    const double dist = plane_vec.dot(point) + w_scaled_padded - 1e-6;
```
Can you replace the hard-coded 1e-6 by detail::ZERO or should it be significantly larger?

peci1 replied:
Tests pass even with detail::ZERO, so I'll use that one.

rhaschke commented on src/bodies.cpp:
```diff
@@ -1142,30 +1181,32 @@ bool bodies::ConvexMesh::intersectsRay(const Eigen::Vector3d& origin, const Eige
   // transform the ray into the coordinate frame of the mesh
   Eigen::Vector3d orig(i_pose_ * origin);
-  Eigen::Vector3d dr(i_pose_ * dir);
+  Eigen::Vector3d dr(i_pose_.linear() * dir);
```
Just another weird bug!

rhaschke reviewed (commented):
Dear @peci1,
please adapt multiple consecutive uses of set*() to set*Dirty() after having merged #126. Otherwise, this looks good. Thanks a lot for all your effort on this!

peci1 committed:
Improved documentation.

peci1 committed:
Utilize setDimensionsDirty() to increase performance.

peci1 committed:
Vectorized and simplified bodies::Box::intersectsRay().

peci1 commented:
@rhaschke I did all the changes you requested. ABI compatibility with melodic is achieved by the very last commit.

Thanks for the deep insights. The code looks much better now.

peci1 commented:
I also restructured the commit history. Now I consider all the commits valuable and further squashing is not needed.

rhaschke commented on src/bodies.cpp:
```diff
@@ -1021,7 +1031,8 @@ void bodies::ConvexMesh::updateInternalData()
   shapes::Box box_shape(mesh_data_->box_size_.x(), mesh_data_->box_size_.y(), mesh_data_->box_size_.z());
   bounding_box_.setPoseDirty(pose);
-  bounding_box_.setPaddingDirty(padding_);
+  bounding_box_.setPaddingDirty(
+      padding_);  // the padding effect will most likely be smaller, but this is the worst case
```
Please improve the comment (if it's really needed).
```suggestion
  // the padding effect will most likely be smaller, but this is the worst case
  bounding_box_.setPaddingDirty(padding_);  
```

rhaschke commented on src/bodies.cpp:
```diff
@@ -1239,13 +1258,29 @@ bool bodies::ConvexMesh::intersectsRay(const Eigen::Vector3d& origin, const Eige
+    {
+      if (intersections->size() == n)
+        break;
+      if (!intersections->empty() && (p.pt - intersections->back()).squaredNorm() < pow(detail::ZERO, 2))
```
candidate for Eigen::isApprox again.

rhaschke commented on src/bodies.cpp:
```diff
@@ -514,8 +520,18 @@ bool bodies::Cylinder::intersectsRay(const Eigen::Vector3d& origin, const Eigen:
+  // If a ray hits exactly the boundary between a side and a base, it is reported twice.
+  // We want to only return the intersection once; the loop makes use of the fact that ipts is
+  // already sorted by distance, so if some points are duplicate, they are examined successively.
+  for (const auto& p : ipts)
+  {
+    if (intersections->size() == n)
+      break;
+    if (!intersections->empty() && p.pt.isApprox(intersections->back(), detail::ZERO))
+      continue;
+    intersections->push_back(p.pt);
+  }
 
```
This code snippet is used twice (see above).
What about introducing an inline function `detail::filter_duplicates(ipts, intersections)` for it?

rhaschke reviewed (commented):
Some minor remarks left.
As I will not cherry-pick to noetic, but revert the last commit, please remove the comment 
`Do not cherry-pick this commit to noetic.` from it.
Thanks a lot for all this work. Hope, we can merge this later today.

peci1 committed:
Refactored intersection code.

peci1 committed:
ABI compatibility for melodic.

peci1 commented:
@rhaschke I addressed your comments.

rhaschke commented:
@peci1, thank you very much. All your PRs are a great improvement of this code base!
If I may ask for even more, here are my suggested next steps:
- use of set*Dirty() functions (for Melodic)
- API cleanup (for Noetic)
  - Consider moving non-trivial functions from `.h` to `.cpp`.
Usually, only one- or two-liners deserve inlining.
  - `min_corner_` and `max_corner_` as suggested in https://github.com/ros-planning/geometric_shapes/pull/109#discussion_r415056537
  - pimpl pattern for MeshData as suggested in https://github.com/ros-planning/geometric_shapes/pull/109#discussion_r415099445
- any other improvements you can think of :wink:

rhaschke commented:
I added a `noetic-devel` branch, just in case you need it ...
